Round 2: CI/CD deployment guide (#272)
gruntwork-io/gruntwork-io.github.io: _posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

allowing you to improve the design incrementally. Similarly, by deploying the software more frequently to production,
the Continuous Delivery process enables you to keep software packages small enough to reduce the risk and impact of each
deployment.

While CI/CD for application code is well understood in the software industry, CI/CD for infrastructure code is a
nascent practice. This guide focuses on providing an overview of the background info, design, and implementation
of a production-ready CI/CD pipeline for infrastructure code.

=== What you'll learn in this guide

This guide consists of four main sections:

<<core_concepts>>::
  An overview of the core concepts you need to understand what a typical CI/CD pipeline entails for infrastructure code,
  including a comparison with CI/CD for application code, a sample workflow, infrastructure to support CI/CD, and threat
  models to consider to protect your infrastructure.

<<production_grade_design>>::
  An overview of how to configure a secure, scalable, and robust CI/CD workflow that you can rely on for your
  production application and infrastructure code. To get a sense of what production-grade means, check out

>   link:/guides/foundations/how-to-use-gruntwork-infrastructure-as-code-library#production_grade_infra_checklist[The production-grade infrastructure checklist].

Author: This will actually work once it deploys, as the URLs are constructed with jekyll.

  integrate and deploy code for a feature without enabling it. By doing so, you are able to continuously integrate new
  developments over time. This guide will briefly introduce automated testing and feature toggles for infrastructure
  code, but will not do a deep dive on the subject. You can learn more about best practices for automated testing in our
  talk
  https://blog.gruntwork.io/new-talk-automated-testing-for-terraform-docker-packer-kubernetes-and-more-cba312171aa6[Automated
  testing for Terraform, Docker, Packer, Kubernetes, and More] and blog post
  https://www.ybrikman.com/writing/2016/02/14/agility-requires-safety/[Agility requires safety].

[[core_concepts]]
== Core Concepts

[[why_is_it_important_to_have_cicd]]
=== Why is it important to have CI/CD?

To understand the benefits of CI/CD, it is worth exploring the opposite: _late integration and late delivery (LI/LD)_.
We will explain LI/LD using a thought experiment about building the International Space Station (ISS).

.Components of the International Space Station
image::/assets/img/guides/infrastructure-cicd-pipeline/iss-components.png[]

> The ISS consists of dozens of components, as shown in the image above. Each component is built by a team from a

Author: I removed the caption as it doesn't add too much value given the figure title in the image itself.

> The ISS consists of dozens of components, as shown in the image above. Each component is built by a team from a

Author: 6a1ff2a32079bc29ea5079308d95d44c517885b9

toilets from the scope thinking some other team is sure to include it. Finding all of this out once everything has
already been built and is floating in outer space means that fixing the problems will be very difficult and expensive.

While it is hard to imagine that anyone would build the ISS in this way, unfortunately this model of development is
fairly common in the software industry. Developers work in total isolation for weeks or months at a time on _feature
branches_ without integrating their work with other teams, and then try to merge all the work together at the last
minute moments before release. As a result, the integration process is very expensive and takes days or weeks fixing merge
conflicts, tracking down subtle bugs, and trying to stabilize release branches.

.Many teams employ a practice of working on their features over long periods of time on isolated branches. These long lived feature branches have higher chances of having big conflicts to resolve when integrated.
image::/assets/img/guides/infrastructure-cicd-pipeline/feature-branch-merge-conflict.png[]

> In contrast, the Continuous Integration and Continuous Delivery model of development promotes more cross team

Author: In the asciidoc rendered view (on the website) it will look ok. See how the caption is rendered here:

<img width="787" alt="How_to_deploy_a_production-grade_Kubernetes_cluster_on_AWS_-_Mozilla_Firefox" src="https://user-images.githubusercontent.com/430092/76339997-17bd2e00-62b8-11ea-97b2-1d8aa4aec5ea.png">

code that is available. There are two distinct types of infrastructure code:

Infrastructure Modules::
  Modules are bundles of infrastructure code that can be used to deploy a specific component of your architecture.
  For example, many companies have modules for deploying private networks using Virtual Private Clouds (VPCs),
  databases, docker clusters (e.g Elastic Container Service, Kubernetes, Nomad), etc. Think of modules as the
  "blueprints" that define the way your company configures infrastructure.

Live Infrastructure Configurations::
  Live infrastructure configurations are specific parameters for each component in your architecture. The live
  configurations are the frontend for your infrastructure deployments. For example, you might define your dev
  environment as a series of configuration files for the modules that specify the various parameters specific to
  development (e.g sizing small instances, naming them with `dev`, using accounts that are accessible by all developers,
  etc). If the modules are "blueprints" then the live configuration contain the "houses" that were built using the

>   "blueprints."

Author: Good idea!

>   "blueprints."

Author: 0fde26a0a5ed26c84899e2bd61752416d5c5a791

Typically you would have separate repositories for each of these (e.g `infrastructure-modules` for modules and
`infrastructure-live` for live configuration). Organizing your infrastructure code in this way ensures that you can
better test the code before they are used to deploy your environments.

> Isolating modules from live configuration allows you to:

Author: Ok. I guess it is fairly clear that these properties come out of the code organization without much explanation.

> Isolating modules from live configuration allows you to:

Author: 4995cd6f54fce35fd1231b68aa02d21dab059162

  reusing the modules across each environment by invoking them with different inputs that are specific to the
  environment.

There are distinct differences in the way the code is tested, used, and deployed between the two flavors of
infrastructure code. These differences are important to consider when designing CI/CD workflows, as they lead to many
differences in the implementation of the pipeline. In the next seciton, we will walk through a typical CI/CD workflow
and compare and contrast the pipeline between the three flavors of code we've talked about so far: application code,
infrastructure modules, and live infrastructure configuration.

[[cicd_workflows]]
=== CI/CD workflows

Now that we have gone over what, why, and how CI/CD works, let's take a look at a more concrete example walking through
the workflow.

The following covers the steps of a typical CI/CD workflow. Any CI/CD workflow, whether it be for infrastructure code or
application code, generally follow these steps. However, the details of some of the steps differ widely. We'll go over
each step of the process and then highlight the differences in implementation for the three flavors of code we covered:

> application code, infrastructure modules, and live infrastructure configuration.

Author: Roger: 61a8a1b

For the purposes of illustrating this workflow, we will assume the following:

- The code lives in version control.
- We are using a trunk-based development model.
- The code has already been in development for a while and there is a version running in production.

Here are the steps:

. <<clone_a_copy_of_the_source_code>>
. <<run_the_code_locally>>
. <<make_code_changes>>
. <<submit_changes_for_review>>
. <<run_automated_tests>>
. <<merge_and_release>>
. <<deploy>>

[[clone_a_copy_of_the_source_code]]
==== Clone a copy of the source code and create a new branch

Typically the first step in making changes to any code base is to clone the repository locally and begin development on
a new branch. Having a local copy makes it easier to iterate on the changes, and using an isolated branch allows you to
push code back to the central repository without breaking the main line of code (trunk) that everyone else is working
on.

> It is important to keep the trunk clean in a trunk-based development model, as all new work starts from the trunk.

Author: I didn't think so at first, and then double checked the trunk-based development section and indeed you are right that we cover each of these already. The perils of working on the same document over multiple weeks.

Removed: 6fbc32d

  way to have a true and complete local copy of a cloud. Therefore, the only way to know for sure your infrastructure
  code works is by making the actual API calls to the cloud to deploy it. With infrastructure modules, this involves
  deploying the module into a sandbox environment. For example, to test a terraform module, you can define example code
  that sets up the necessary resource dependencies that the module needs, and then deploy that into your sandbox with
  `terraform apply`. You can then inspect the deployed resources to make sure they are functioning as expected. For
  convenience, this process could be captured in an automated test using a framework such as
  https://terratest.gruntwork.io/[Terratest].

Live Infrastructure Config::
  Locally testing live infrastructure config is more difficult than either application code or infrastructure modules.
  Unlike with infrastructure modules, it is difficult to deploy the live infrastructure config temporarily as the code
  is tied to a specific live environment by nature of the code. After all, this is the configuration to manage live

>   infrastructure. To illustrate this point, consider a scenario where you are working on updating the cross account IAM

Author: Done: f50948a

  roles to access your environments, and you are at the point of reflecting your changes to prod. Would you want to
  deploy that code to your live production environment off of an unreviewed branch? +

>   The only real test you can do for live infrastructure config is to do a dry run of your infrastructure code. Most Infrastructure as Code tools support a dry run of the code to check what it would do against your environment. For example, with Terraform, you could run `terraform plan` to sanity check the planned actions Terraform will take. This is especially useful for sanity checking a fresh clone of the code. The trunk should be a true reflection of the live environment, so you should expect there to be no changes to make on a fresh clone of trunk.

Author: I tried a few ways, but in the end could not figure out a way to do this in the definition list. I think I need to switch from definition lists, but then it makes it hard to do the switch over from side-by-side-comp => overall explanation at the end.

I think I am going to keep as is for now, and then revisit later, since I am already overbudget on this project and guide.

a dry run). However, testing infrastructure modules can take a long time since you need to deploy infrastructure (on the
order of 10s of minutes). For ideas on how to improve the test cycles for infrastructure modules, take a look at
https://terratest.gruntwork.io/docs/testing-best-practices/iterating-locally-using-test-stages/[Iterating locally using
test stages] in the Terratest documentation.

[[submit_changes_for_review]]
==== Submit changes for review

Once the code implementation is done and the testing passes, the next step is to submit it for review. You want to focus
your review process on things that are hard to check through automated testing, such as checking security flaws,
reviewing general code design, enforcing style guides, or identifying potential performance issues on larger data sets.
Code review processes are also a great way to share knowledge across the team. The reviewer will oftentimes share
valuable insights on the code that you might not have thought of.

How thorough the review should be depends on the amount of automated testing. If you have a comprehensive test suite

> that everyone on the team has confidence in, then it is not necessary to have a thorough review process. On the other

Author: > These are hard things to test for that a thorough and well executed code review will highlight.

I was trying to get at the fact that infrastructure code requires an even more thorough review to look for and account for bugs because you don't have automated testing, but you are right that I probably went too overboard in the claim here.

Removed as Jim suggested to avoid the politics: 9e42e51

hand, if you have no automated tests or working on an area that can't be tested automatically (e.g live infrastructure
code), then you'll want as many eyes as possible reviewing the code.

[[run_automated_tests]]
==== Run automated tests

To help with code review, you should also set up a CI server (such as Jenkins or CircleCI) with commit hooks that
automatically trigger testing of any branch that is submitted for review. Running the automated tests in this fashion

> not only ensures that the code passes all the tests, but also verifies that it runs on multiple platforms and not just

Author: Reworded: f985644

  make sure the artifact is immutable (i.e., you never change it), and that it has a unique version number (so you can
  distinguish this artifact from all the others).

Infrastructure Modules::
  Infrastructure modules are typically consumed as a library in the tool. Most infrastructure as code tools consume
  libraries directly from a Git repository. For example, with Terraform you can consume modules through module blocks
  that reference a Git repository (see
  https://www.terraform.io/docs/configuration/modules.html[the official documentation] for more details). In this case,
  using a Git tag to mark a revision with a human friendly name is sufficient to generate the release artifact.

Live Infrastructure Config::
  For live infrastructure config, there is typically no release artifact. Live infrastructure code doesn't need to be
  packaged to deploy as it is directly consumable. For example, for Terraform or Terragrunt live config, you can
  directly run `terraform apply` or `terragrunt apply` on the repo. In general, it is not necessary to tag your commits
  for live infrastructure config because in practice you will end up deploying every commit of trunk.

> It is worth expanding a bit on the reason why live infrastructure config does not have any release artifact. To

Author: I want to keep this section. This was a major source of confusion for Ben, and I suspect for a large number of people as well.

new changes to make to the live environments when you start. This is so that you get an accurate representation of the
changes that are being introduced, since you don't want to be differentiating between existing changes that will be
applied from trunk and the changes that will be applied with your new code.

Given that, the definition of a "clean build" for the trunk with live infrastructure config is that a dry run returns no
changes to make. This in turn means that the latest state of trunk that you are working off of should be a
representation of what is actually deployed in your environments. Therefore, to ensure the trunk is clean, you will need
to make sure that you continuously deploy and apply the trunk as new code is merged in.

This leads to what we call _The Golden Rule of Infrastructure Code:_

*_The master branch of the live repository should be a 1:1 representation of what's actually deployed._*

You will want to do everything that is in your power to maintain this representation to streamline your development.

[[deploy]]
==== Deploy

Now that you have a release artifact, the final stage of the process is to deploy the code. The act of deploying the

> code immediately after generating the release is Continuous Deployment (CD).

Author: Ah I didn't know that distinction. I am also not sure a lot of people know that distinction either. Do you think it is worth describing that in this document somewhere?

In any case, I am going to remove that sentence to avoid the accuracy problem and avoid definitions of terms, since it is really just fluff.

> code immediately after generating the release is Continuous Deployment (CD).

Reviewer: It's from the book *Continuous Delivery**, IIRC. No need to define it here, just want to be sure we use the right terms.

  image. Regardless of how your application is deployed, it is important to reflect the changes in your live
  infrastructure config to perform the deployment. Note that there are various strategies for deploying application
  code, such as canary and blue-green deployments. We will not get into details here, but you can refer to our post
  https://blog.gruntwork.io/how-to-use-terraform-as-a-team-251bc1104973#7dd3[How to use Terraform as a team] for an
  overview of various rollout strategies.

Infrastructure Modules::
  To deploy your infrastructure modules, you need to create or update references to the modules in your live
  infrastructure config. If the module is already deployed, this may be as simple as bumping the ref tag in your live
  config. However, if the module is being deployed for the first time, then this will require creating a new
  configuration in your live infrastructure config to deploy the module. In either case, the only way to deploy
  infrastructure modules is by making the corresponding edits to the live infrastructure config to roll out the changes

>   across your environments. Note that because infrastructure changes are higher risk than application code, you will

Author: Makes sense. Removed: aa4c2ec

  want to make sure to roll out your infrastructure in stages. That is, you should fully roll out the changes in a
  pre-prod environment before moving on to deploying in prod.

Live Infrastructure Config::
  For live infrastructure config, deploying the code is the act of applying the code to the live environment. This
  depends on the tool. For example, your terraform code can be applied with `terraform apply` or `terragrunt apply`,
  while Kubernetes manifests require `kubectl apply`.

Regardless of the code, the important thing about the deployment process is that you should automate as many steps of
the process as possible. Just like with automated testing on code review, you should set up a CI server to automatically
deploy your code across the various environments, to the extent that it is safe. You can automate many steps of the

> deployment depending on the risk profile of the change:

Author: Ok. Done: ebf3ce7

Application Code::
  For application code, you should be able to automate the entire deployment as the surface area of each change should
  be fairly small and localized to just the application. Additionally, application code have standard rollback
  practices that make it easy to undo. Automating the deployment of application code requires automatically patching

>   your live infrastructure config. In most cases, the changes are simple enough that a single `sed` call should

Author: This section was consolidated and removed: ebf3ce7

  be sufficient (e.g bumping the version tag of a docker image). Note that the roll out of the deployment should be
  handled as part of the automated deployment of the live infrastructure config.

Infrastructure Modules::

>   With infrastructure modules, depending on which module is being updated, an automated deployment may be risky. In a

Author: Ah that is a good point. Updated the wording: e4e9e60

  world where a single name change could cause the database to be recreated, it pays off to be extra cautious with the
  deployment. At most, you should only automatically deploy infrastructure modules to a sandbox or dev environment,
  where the damage from a catastrophic failure causing the environment to be taken down is limited. Similar to

>   application code, automating this process could be handled with a single `sed` call that updates the ref tag in the

Author: This section was consolidated and removed: ebf3ce7

* Many strategies: canary, blue-green, rolling deployment.
* Promote immutable, versioned artifacts across environments.

|
* Manually update Live Infrastructure Config with new ref tag.
* Limited deployment strategies.
* Promote immutable, versioned artifacts across environments.

|
* Continuously deploy directly from master (with approval workflow).
* Only one deployment strategy.

|===

The rest of the document will discuss how we can implement the automated pieces of the workflow in a secure manner that
is ready for production.

To start, let's take a step back and define a threat model for CI/CD. This threat model will help us ensure that we
implement the necessary security controls in these CI/CD pipelines so that we cover the common types of attack vectors
for this type of workflow.

[[threat_model_of_cicd]]
=== Threat model of CI/CD

The threat model of CI/CD is different between application code, infrastructure modules, and live infrastructure config.
This largely stems from the amount of permissions required to implement each workflow:

.Permissions needed for CI/CD of each code type.
[cols="1h,2,2,2"]
|===

> |Permissions |Application Code |Infrastructure Modules |Live Infrastructure Config

Author: Removed: ab4b6e7

|CI server logs
|Limited (typically only need summary report as you can reproduce failures locally)
|Full (can not debug without seeing the full deployment log)
|Full (can not debug without seeing the full deployment log)

|Sandbox environments
|Limited (minimal necessary to deploy artifact)
|Full
|Full

|Non-prod environments
|Limited (minimal necessary to deploy artifact)
|None
|Full

|Prod environments
|Limited (minimal necessary to deploy artifact)
|None
|Full

|===

Note how CI/CD for infrastructure code requires full permissions to CI logs as well as at least one of your
environments, whereas application code can be deployed with limited permissions to the environments. This difference is
due to the fact that infrastructure modules and live infrastructure config can be used to manage any infrastructure
deployed in your environments, whereas application code is scoped to infrastructure that is relevant to the application.

> Given that, the threat against CI/CD in infrastructure code is orders of magnitude higher than with application code.

Author: Removed this whole section and made it more generic: ab4b6e7

Application Code::

>   Compromising the credentials used for application code may be used to take down your application in a denial of service

Author: Removed: ab4b6e7

>   Compromising the credentials used for application code may be used to take down your application in a denial of service

Author: I had a naive assumption where the credentials only give you access to deploy a new version of the app, but then I realized that you can deploy a new app with the same credentials as the original app so you can get access to the app data.

mitigation tactics for those threats. This is where threat modeling helps.

A threat model explicitly covers what attacks are taken into consideration in the design, as well as what attacks are
__not__ considered. The goal of the threat model is to be realistic about the threats that are addressable with the
tools available. By explicitly focusing attention on more likely and realistic threats, we can avoid overengineering and
compromising the usability of the solution against threats that are unlikely to exist (e.g a 5 person startup with 100
end users is unlikely to be the subject of a targeted attack by a government agency).

In this guide, the following threat assumptions are made:

- Attackers' goals are to gain access to an environment that they do not already have access to. Access to an
  environment includes but is not limited to:

    * The ability to read secrets that grant access to potentially sensitive data (e.g the database in prod
      environment).
    * Full access over all resources to cause damage to the business (e.g ability to delete the database and all its
      backups in prod).

> - Attackers can originate from both external and internal sources (in relation to the organization).

Author: Agreed! But since this is already long, I'll regrettably keep it as is for now instead of honing the point further.

- External attacks are limited to those that can get full access to a CI environment, but not the underlying source
  code. Note that __any__ CI/CD solution can likely be compromised if an attacker has access to your source code.
- Internal attackers are limited to those with restricted access to the environments. This means that the threat model
  does not consider highly trusted insiders who abuse their privileges with malicious intent (e.g
  internal ops admin with full access to the prod environment). However, an internal attacker with permissions in the
  dev environment trying to elevate their access to the prod environment is considered.
- Similarly, internal attackers are limited to those with restricted access in the CI environment and git repository. A
  threat where the internal attackers can bypass admin approval in a CI pipeline or can force push deployment branches
  is not considered.
- Internal attackers can have (limited) access to the CI environment and the underlying code of the infrastructure (e.g
  the git repository).

With this threat model in mind, let's take a look at the different CI/CD platforms.

[[cicd_platforms]]

> === CI/CD platforms

Author: This is painful for me since I really like the text overview. The table is very terse and it leaves a lot of room for questioning since there is no explanation. But... I guess you are right about shortening the guide and see how people receive this.

Consolidated in: 6c6d2df

  Most vulnerabilities are mitigated by locking down the platform so that it is only internally accessible within a
  corporate network over VPN, as well as by keeping the server, the Jenkins software, and the Jenkins plugins frequently patched. This does mean that you shouldn't
  expose the server externally to implement webhook based workflows, but rather rely on polling to detect changes.
  Jenkins also provides a wide range of fine-grained permissions in its user model that allows you to lock down the
  actions that your internal users can take.

https://circleci.com/[CircleCI] (SaaS, generic)::
  CircleCI is a generic SaaS CI/CD platform that is more optimized for CI/CD workflows for application code than
  infrastructure CI/CD, although it can be configured to implement infrastructure CI/CD workflows. It has first class
  support for various application based workflows like parsing test results into a nice report in the UI, reacting to
  version control events, and official docker containers for various programming languages and frameworks. However, it
  does not have first class support for features that are useful with infrastructure code, such as encrypting logs

>   and fine grained permissions. The strength of CircleCI lies in the simplicity of its infrastructure. Being a fully

Author: This section was removed and consolidated to just the table in the interest of shortening the document. 6c6d2df

  deployment policies and permissions as code. Note that Terraform Cloud and Enterprise are both optimized for running
  `terraform`. Both do not support directly calling external binaries such as `terragrunt` or `docker` in a workflow.
  Terraform Cloud in particular does not support the use of external binaries even when called from within `terraform`
  using a `local-exec` provisioner or external data source. In terms of the threat model, Terraform Cloud and Enterprise
  both implement respectable security practices, including end to end encryption backed by HashiCorp Vault, and you have
  the option to run in your own data center with Terraform Enterprise to lock it down. Note that Terraform Enterprise
  relies on API based integrations with the VCS platforms instead of webhooks, so it is not necessary to have a publicly
  facing service when running in self-hosted mode.

Here is a table that highlights some of the features of each CI platform and their strengths and weaknesses that can be
useful in making tradeoffs:

[cols="1h,2,2,2,2,2"]
|===
| |Jenkins |CircleCI |GitLab |Atlantis |TFE and TFC

|Hosting
|Self-hosted
|SaaS
|SaaS or Self-hosted
|Self-hosted

> |SaaS or Self-hosted

Author: Updated table with your info and added in buildkite: 6c6d2df

|Generic?
|Yes
|Yes
|Yes
|Specialized to terraform
|Specialized to terraform

|VCS integration
|Yes, with plugins
|Yes
|Yes
|Yes
|Yes

|Provides static IP addresses for IP whitelisting
|Yes
|No
|Yes (self-hosted)
|Yes
|Yes (TFE)

|Update commit statuses
|Yes
|Yes
|Yes
|Yes
|Yes

|Annotate pull requests
|Requires custom scripting
|Requires custom scripting
|Requires custom scripting
|Yes
|Yes

|Supports multiple infrastructure tools
|Yes
|Yes
|Yes
|Limited (terraform only; additional binaries can be installed, but can not be called directly)
|No (terraform only; TFE can support additional binaries, but TFC does not)

|===

[[production_grade_design]]
== Production-grade design

With all the core concepts out of the way, let's now discuss how to configure a production-grade CI/CD workflow for
infrastructure code, using a platform that looks something like this:

> .Architecture of platform for running Terraform/Terragrunt CI/CD workflows.

Reviewer 2: I found this diagram to be a bit confusing.

First, in the CI Servers box, the CI icons on the left side in the corner and then multiple raised boxes didn't make sense to me. Perhaps just have one raised box labeled "CI server" and leave off the icons?

Next, why are there multiple Lambda functions, but an arrow to just one of them? Also, the "Invoke Lambda Functions" text is too small to read. 

In the ECS box, I'd leave off the "ECS Fargate" icon, then rotate the "ECS Fargate" text 90deg and increase font size.

Reviewer 2: I actually wonder if CloudCraft is the best tool here. A flat 2D diagram might be more effective.

Author: Ok I updated the image given your feedback: 93034c5

What do you think?

> Perhaps just have one raised box labeled "CI server" and leave off the icons?
IMO, the icons are important because they give a visual cue as to what this CI server should be.

image::/assets/img/guides/infrastructure-cicd-pipeline/tftg-pipeline-architecture.png[]

[[use_generic_cicd_platforms_as_a_workflow_engine_but_run_infrastructure_deployments_from_within_your_account]]
=== Use generic CI/CD platforms as a workflow engine but run infrastructure deployments from within your account

Given the limitations and tradeoffs of the various platforms we covered in <<cicd_platforms>>, we don't recommend
relying on a single platform for implementing the entire workflow. Instead, we recommend a hybrid solution that takes
advantage of the strengths of each platform, and cover the weaknesses. The design looks as follows:

> - Use any generic CI/CD server you wish to integrate with your VCS repositories.

Reviewer 2: "This could be Jenkins, CircleCI, GitLab, or whatever you prefer."

> - Use any generic CI/CD server you wish to integrate with your VCS repositories.

Author: Took Jim's suggestion and applied it: 73cbdea

> - Use any generic CI/CD server you wish to integrate with your VCS repositories.

Reviewer: Whoops, had a typo in my description that I think got copied into the doc:

> that can be used to define pre-defined commands

Should be:

> that can be used to run pre-defined commands

> - Use any generic CI/CD server you wish to integrate with your VCS repositories.

Author: Fixed https://github.com/gruntwork-io/gruntwork-io.github.io/pull/273

  a limited trigger we'll define below.
- Define your CI workflows so that the CI/CD server triggers deployments against the deploy server.

This design implements separation of the concerns so that we take full advantage of the strengths of each platform,
while covering the weaknesses: relying on the CI/CD platforms to manage the workflow/pipeline, but have it trigger
infrastructure deployments on self-hosted systems that are more locked down.

We don't want to give the CI/CD servers permissions to deploy and manage arbitrary infrastructure. CI/CD servers are
typically not secure enough to handle sensitive information, and you don't want a server that is used for executing
arbitrary code and regularly used (and written to) by your entire dev team to have admin permissions.

Instead, we delegate this responsibility to an isolated, closed off system in the AWS account that only expose a limited
set of actions that can be triggered. That way, if anyone gets access to your CI server, they can at most kick off
builds on existing code, but they don't get arbitrary admin access.

[[options_for_deploy_server]]

> === Options for deploy server

Reviewer 2: Observation: This section directly contrasts Gruntwork ECS Deploy Runner Stack with Terraform Enterprise, and paints Gruntwork's solution in a better light. Them are fightin' words.

Author: That was not the intention, so highlighted the strengths of TFE better: 0093729

What do you think of the updated wording?

The deploy server needs to be a self-hosted platform in order to satisfy the requirement for isolation. It should also
avoid executing arbitrary workflows. Finally, it should support configurations options that limit what code can run on
the server. This limits the options for what you can use as your deploy server. Here is a list of platforms that satisfy
these constraints, and their strengths and weaknesses:

Gruntwork ECS Deploy Runner Stack::
  This is a stack you can deploy in your AWS account that sets up an ECS task with a customizable docker container for
  running `terraform validate`, `terraform plan`, and `terraform apply`, or the Terragrunt equivalent. It is also

>   extensible to support other commands as well, such as running `go test` for Terratest or `packer build` for building

Reviewer 2: Perhaps make it clear that it can be extended to run *any* arbitrary command!

Author: I think I will keep as is, since that could be seen as a security risk if someone glosses over it.

  images. To limit the ability to run arbitrary code, the stack includes a Lambda function that can be used as a trigger
  which exposes a limited set of options and additional checks for source repository. It relies on serverless
  technologies to limit the amount of overhead required for maintaining the system.

> Terraform Enterprise::

Author: TFE has the sentinel policies though, which I think would be super powerful in CI/CD.

> Terraform Enterprise::

Author: Updated the wording to clarify this point: 0093729

What do you think?

> Terraform Enterprise::

Reviewer: I think that clarifies some things, but doesn't quite touch on the fact that if you're allowing devs to click around TFE, provide various input variables in the UI, and kick off deployments that leverage TFE's admin permissions, some of the security benefits discussed in this doc don't quite work.

> Terraform Enterprise::

Author: I feel like using TFE as a deploy server in our workflow is orthogonal to using TFE for self-service. For example, you could use TFE just for the sentinel purpose when driving deployments in CI/CD fashion and close off the UI. Sure that lowers the value of TFE, but it doesn't make it less secure than the ECS deploy runner. And you get sentinel policies with it.

So there are extra steps one needs to take to use TFE as a deploy server (this is why being super detailed about the threat model here is important!), but it doesn't make TFE completely incompatible. Compare this with Atlantis, where you can't manually trigger deployments, or TFC where it is hosted and thus you have to share your credentials with the third party. There is no way to use those tools in the workflow we described while mitigating the threats, and hence are completely incompatible.

repository. This means that any public repo can cause your Atlantis server to run `terraform plan` and `terraform apply`
on custom code you do not control using the permissions granted to that server. Instead, you will want to configure it
so that only certain repositories, branches, and users can trigger the workflow.

The Gruntwork ECS Deploy Runner stack mitigates this concern by only allowing triggers from a Lambda function that
exposes a limited set of actions against the deploy runner task. The lambda function:

- Requires a single repository to trigger deployments by default.
- Can be configured to limit deployments to specific branches.
- Requires explicit IAM permissions to trigger.

You can find similar mechanisms for limiting deployments in the various deploy server options.

[[use_a_vpc_to_lock_down_infrastructure_deployer]]
=== Use a VPC to lock down deploy server

Run your infrastructure deployment workloads in a https://aws.amazon.com/vpc/[Virtual Private Cloud (VPC)] to isolate
the workloads in a restricted network topology (see link:/guides/networking/how-to-deploy-production-grade-vpc-aws[How

> to deploy a production-grade VPC on AWS] for more information on VPCs). Configure to run all workloads in private

Reviewer 2: ```suggestion
to deploy a production-grade VPC on AWS] for more information on VPCs). Configure it to run all workloads in private
```

subnets that are not publicly accessible. Make sure to block all inbound internet access and consider blocking all
outbound access except for the minimum required (e.g, allow access to AWS APIs).

[[use_minimal_iam_permissions_for_a_deployment]]
=== Use minimal IAM permissions for a deployment

Avoid having a single system with admin permissions for running a deployment. Instead, deploy specialized versions of
the deployment platforms with varying permissions for handling specific workflows. By separating out the concerns for
each pipeline, you can reduce the blast radius of the damage that can be done with each set of credentials. At a minimum,
you should have two versions of the infrastructure deployment system: one for deploying the application code, and which
only has the minimal permissions necessary for deploying that application; and one for deploying infrastructure code,

> which has full access to the environments.

Reviewer 2: ```suggestion
which has more access to the environments.
```

Even for the infra deployments, it would be ideal if it could have reduced permissions. In fact, to be compliant with CIS, you couldn't have a deployment system with full access.

[[use_approval_flows]]
=== Use approval flows

> It is important that human review is baked into each deployment. As covered in <<cicd_for_infrastructure_code>>, it is

Reviewer 2: `<<cicd_for_infrastructure_code>>` <== does this need to be fixed?

Author: Ah yes. Good catch!

difficult to build an automated test suite that builds enough confidence in your infrastructure code to do the right
thing. This is important, as failed infrastructure deployments could be catastrophic to your business, and there is no
concept of rollback with infrastructure deployment tools. This means that you will almost always want to have some form
of approval workflow for your infrastructure CI/CD pipeline so that you can review what is about to be deployed. Most
generic CI/CD platforms support approval workflows. For example, CircleCI supports
https://circleci.com/docs/2.0/workflows/#holding-a-workflow-for-a-manual-approval[approval steps in its workflow
engine], in addition to https://circleci.com/docs/2.0/contexts/#restricting-a-context[restricted contexts] to limit who
can approve the workflow.

[[lock_down_vcs_systems]]
=== Lock down VCS systems

It is a good practice to define and store the deployment pipeline as code in the same repo that it is used. For example,
you should define the CI/CD deployment pipeline for your infrastructure code in the `modules` and `live` repositories.

> However, this means that anyone with access to those repositories are free to modify the pipeline, __even on feature

Reviewer: ```suggestion
However, this means that anyone with access to those repositories could modify the pipeline, __even on feature
```

branches__. This can be exploited to skip any approval process you have defined in the pipeline by creating a new branch
that overwrites the pipeline configuration.

This is not a concern if only admin users had access to the infrastructure code. Typically, however, many operations
teams want contributions to the infrastructure code from developers as well, and having any developer have the ability to
deploy arbitrary infrastructure to production without any review can be undesirable. To mitigate these concerns, you
should lock down your VCS systems:

Only deploy from protected branches::
  In most git hosting platforms, there is a concept of protected branches (see
  https://help.github.com/en/github/administering-a-repository/about-protected-branches[GitHub docs] for example).
  Protected branches allow you to implement policies for controlling what code can be merged in. For most platforms, you

>   can protect a branch such that: (a) it can never be force pushed, (b) it can never be merged to or commit to from the

Reviewer 2: This list might be more readable on separate lines.

Author: Agreed, but due to the limitations of asciidoc definition lists it's a bit hard to massage this. I'm going to keep it as is for now and we can revisit later.

  cli, (c) merges require status checks to pass, (d) merges require approval from N reviewers. By only building CI
  pipelines from protected branches, you can add checks and balances to ensure a review of potentially harmful
  infrastructure actions.

Consider a forking based workflow for pull requests::
  When exposing your repository to a wider audience for contribution, you can consider implementing a forking based
  workflow. In this model, you only allow your trusted admins to have access to the main infrastructure repo, but anyone
  on the team can read and fork the code. When non-admins want to implement changes, instead of branching from the repo,
  they will fork the repo, implement changes on their fork, and then open a PR from the fork. The advantage of this
  approach is that many CI platforms do not automatically run builds from a fork for security reasons. Instead, admins

>   manually trigger a build by pushing the forked branch to an int

Reviewer 2: As written, the last sentence here is a confusing.

>   manually trigger a build by pushing the forked branch to an int

Reviewer 2: ```suggestion
  secrets, for example. The IAM policies are set up such that the IAM user for CI only has
```

Reviewer 2: I don't understand what github's suggested change is doing here. All I'm suggesting is to drop the phrase "It is important to note that" from the beginning of that sentence.

>   manually trigger a build by pushing the forked branch to an int

Reviewer 2: Consider adding a blank line above this.

>   manually trigger a build by pushing the forked branch to an int

Reviewer 2: Should we use a newer release?

Author: This is actually the latest:

<img width="874" alt="Configuring_CircleCI_-_CircleCI_-_Mozilla_Firefox" src="https://user-images.githubusercontent.com/430092/76347521-067a1e80-62c4-11ea-9e7d-cdf4736e0a48.png">

>   manually trigger a build by pushing the forked branch to an int

Reviewer: I'd say limiting the CI server to read-only access is prob good enough

Author: Updated: d5227b3

>   manually trigger a build by pushing the forked branch to an int

Reviewer: ```suggestion
However, the reality is that __any__ CI/CD solution can likely be compromised if an attacker has full access to your source code.
```

>   manually trigger a build by pushing the forked branch to an int

Reviewer: This part here and the threat model section above goes into a LOT of detail on the security aspects. This is important, but it's also quite long...

Here's an idea: what if we extract the threat model into a separate doc? We could, for example, put it in a Markdown doc in the `module-security` repo and link to that doc from this guide by using the `/infrastructure-as-code-library` page on gruntwork.io. Then, in this guide, we can reduce the security details to a much smaller subset: namely, that deploying arbitrary infra changes requires arbitrary (admin) permissions, and giving a CI server admin permissions is a bad idea as (a) CI servers aren't known to be super secure, (b) you give your whole team access to those servers and let them execute arbitrary code on them, so now everyone on your team is effectively an admin. Our solution mitigates this by giving the admin permissions solely to a separate deploy server. For more details on the threat model, see `<this doc>`.

Author: I know we want to pare down the docs, but I think I am going to keep it as is for now. My main rationale for this is that the threat model and the security aspects are precisely why this isn't as simple as "use Atlantis." So having the doc focus on it and spend a lot of time on it is important IMO.

>   manually trigger a build by pushing the forked branch to an int

Reviewer: Do you need all these snippets and examples? Or could you just go straight into the next section about deploying the ECS deploy runner?

Author: I was just following what you had in the k8s guide. IMO, I think it is good to reiterate at least the folder structure so we know what assumptions we are working with.

With that said, the actual code contents are probably not necessary so removed: d7775d3

>   manually trigger a build by pushing the forked branch to an int

Reviewer: Neat 👍

>   manually trigger a build by pushing the forked branch to an int

Reviewer: I think having these all defined within the guide is good for now, but when we switch to the service catalog model, and are no longer setting up CI / CD for customers, we'll want to create reusable scripts and copy/pasteable code for customers to use to make the setup process easier.

Author: Yup makes sense, although I think more thought is needed in the script design. E.g can we make a drop in file setup?

Reviewer: Def worth thinking through!
